Gems/SceneProcessing/Code/Source/Generation/Components/MeshOptimizer/MeshOptimizerComponent.cpp

```diff
@@ -96,6 +96,85 @@ namespace AZ::SceneGenerationComponents
+    // Vector3 as a key into a unordered_map.
+    template <class MeshDataType>
+    class Vector3Map
+        : private AZStd::unordered_map<AZ::Vector3, AZ::u32>
+    {
+    public:
+        Vector3Map(const MeshDataType* meshData, bool hasBlendShapes, float positionTolerance)
+            : m_meshData(meshData)
+            , m_hasBlendShapes(hasBlendShapes)
+            , m_positionTolerance(positionTolerance)
+            , m_positionToleranceReciprocal(1.0f / positionTolerance)
+        {
+        }
+
+        using AZStd::unordered_map<AZ::Vector3, AZ::u32>::reserve;
+
+        AZ::u32 operator[](const AZ::u32 vertexIndex)
+        {
+            if (m_hasBlendShapes)
+            {
+                // Don't attempt to weld similar vertices if there's blendshapes
+                // Welding the vertices here based on position could cause the vertices of a base shape to be welded,
+                // and the vertices of the blendshape to not be welded, resulting in a vertex count mismatch between
+                // the two
+                return m_meshData->GetUsedPointIndexForControlPoint(m_meshData->GetControlPointIndex(vertexIndex));
+            }
+
+            const auto& [iter, didInsert] = try_emplace(GetPositionForIndex(vertexIndex), m_currentOriginalVertexIndex);
+            if (didInsert)
+            {
+                ++m_currentOriginalVertexIndex;
+            }
+            return iter->second;
+        }
+
+        [[nodiscard]] AZ::u32 at(const AZ::u32 vertexIndex) const
+        {
+            if (m_hasBlendShapes)
+            {
+                // Don't attempt to weld similar vertices if there's blendshapes
+                // Welding the vertices here based on position could cause the vertices of a base shape to be welded,
+                // and the vertices of the blendshape to not be welded, resulting in a vertex count mismatch between
+                // the two
+                return m_meshData->GetUsedPointIndexForControlPoint(m_meshData->GetControlPointIndex(vertexIndex));
+            }
+
+            auto iter = find(GetPositionForIndex(vertexIndex));
+            AZSTD_CONTAINER_ASSERT(iter != end(), "Element with key is not present");
+            return iter->second;
+        }
+
+    private:
+
+        AZ::Vector3 GetPositionForIndex(const AZ::u32 vertexIndex) const
+        {
+            // Round the vertex position so that a float comparison can be made with entires in the map
+            // pos = floor( x * 10 + 0.5) * 0.1
+            return AZ::Vector3(
+                AZ::Simd::Vec3::Floor(
+                    (m_meshData->GetPosition(vertexIndex) * m_positionToleranceReciprocal + AZ::Vector3(0.5f)).GetSimdValue()
+                )
+            ) * m_positionTolerance;
+        }
+
+        const MeshDataType* m_meshData;
+        bool m_hasBlendShapes;
+        float m_positionTolerance;
+        float m_positionToleranceReciprocal;
+        AZ::u32 m_currentOriginalVertexIndex{};
```

Comment:
Can we please use "= 0;" Here? Even if it does the same, = 0 is more explicit and less cryptic.

Reply:
Done.